api/rql/internal/predicate/expression/not.go

```diff
@@ -44,32 +44,16 @@ func (n *not) Unmarshal(input interface{}) error {
 	return nil
 }
 
-// INVARIANT: n.p is an atom by the time each of these Eval*
-// methods are called. This only matters for EvalEntry,
-// EvalEntrySchema and EvalValue.
-
 func (n *not) EvalEntry(e rql.Entry) bool {
-	a := n.p.(*atom)
-	result := a.p.(rql.Primary).EntryInDomain(e)
-	if ep, ok := a.p.(rql.EntryPredicate); ok {
-		result = result && !ep.EvalEntry(e)
-	}
-	return result
+	return !n.p.(rql.EntryPredicate).EvalEntry(e)
 }
 
 func (n *not) EvalEntrySchema(s *rql.EntrySchema) bool {
```

Comment:
This is still a bit of an issue b/c something like `["NOT", ["path", <string_p>]]` would cause the SchemaP to return false. Again, we could document that negation only makes sense for the BooleanPredicate, and to not negate anything else. Could also enforce this in the AST by defining PEs as being made up of only AND/OR, and NPEs (negatable PE) as being made up of AND/OR/NOT. Then the RQL could be something like

```
Query := PE Primary
...
NamePrimary => ["name", NPE StringPredicate]
...
```

Not sure which approach is best. Keeping everything as PEs results in simpler code, but some (possibly) annoying edge cases/debugging. SO maybe NPEs are a good idea.

Reply:
Annoying edge cases/debugging are also code, so probably debatable whether it's simpler. NPE sounds good for clarity.

Will `["NOT", ["path", <string_p>]]` now cause an error?

Reply:
Yes it will. But `["path", [NOT, <string_p>]]` wont. And that kind of makes sense, you are querying the "path" field and that has a predicate assigned. So NPE seems like the right direction.

Reply:
I'm also wondering whether the `true/false` primaries are necessary.. Can't really think of a good reason to include them. I kind of wonder if they were there b/c they made it easy to test the operators back in the old days, where writing unit tests was a lot harder.

Reply:
`true`/`false` definitely make sense as value predicates though.

Reply:
So another thing that's nice about this is negation matches user intent. So something like `["NOT", ["object", [["key", "foo"], p]]` reads "return true if NOT an object w/ key foo whose value satisfies p", while `["object", [["key", "foo"], ["NOT, p]]]` reads "return true if obj w/ key foo whose value satisfies NOT(p)".

Similarly, `["NOT", ["number, p]]` vs. `["number", [NOT, p]]` -- former reads "return true if _not_ number" satisfying p, other is "return true if number satisying NOT p".

`wash find` will probably still need its custom negation semantics, but that's b/c we're assuming user intent is always to trickle-down the NOT. Different front ends can change their parsing rules to accommodate different user intent, but the RQL's user intent is still generic.

Reply:
I think reducing the number of ways to express the same thing makes sense.

Reply:
Meh, may also be good enough to keep `wash find` symmetric with RQL once that refactor's done.